core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java

```diff
@@ -145,21 +140,30 @@ private static void putReportedMetrics(IndexedSet<Metric> metricSet, String inst
         // FULL_NAME_INDEX is a unique index, so getFirstByField will return the same results as
         // getByField
         Metric oldMetric = metricSet.getFirstByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        double oldVal = oldMetric == null ? 0.0 : oldMetric.getValue();
-        Metric newMetric = new Metric(metric.getInstanceType(), metric.getHostname(),
-            metric.getMetricType(), metric.getName(), oldVal + metric.getValue());
-        for (Map.Entry<String, String> tag : metric.getTags().entrySet()) {
-          newMetric.addTag(tag.getKey(), tag.getValue());
+        if (oldMetric != null) {
+          metric.addValue(metric.getValue());
+        } else {
+          Metric newMetric = new Metric(metric.getInstanceType(), metric.getHostname(),
```

Comment:
Why are the metrics updated by creating a new object for every one? That seems very costly...

I feel like we should be able to make Metric atomic and concurrent but using atomic data structures.